pageserver/src/object_repository.rs

```diff
@@ -0,0 +1,909 @@
+            ancestor_lsn: self.ancestor_lsn,
+        })
+    }
+
+}
+
+
+
+///
+/// We store two kinds of page versions in the repository:
+///
+/// 1. Ready-made images of the block
+/// 2. WAL records, to be applied on top of the "previous" entry
+///
+/// Some WAL records will initialize the page from scratch. For such records,
+/// the 'will_init' flag is set. They don't need the previous page image before
+/// applying. The 'will_init' flag is set for records containing a full-page image,
+/// and for records with the BKPBLOCK_WILL_INIT flag. These differ from PageImages
+/// stored directly in the cache entry in that you still need to run the WAL redo
+/// routine to generate the page image.
+///
+#[derive(Debug, Clone)]
+#[derive(Serialize, Deserialize)]
+enum PageEntry {
+    Page(Bytes),
+    WALRecord(WALRecord),
+}
+
+///
+/// In addition to page versions, we store relation size as a separate, versioned,
+/// object.
+///
+#[derive(Debug, Clone)]
+#[derive(Serialize, Deserialize)]
+pub enum RelationSizeEntry {
+    Size(u32),
+
+    // Tombstone for a dropped relation.
+    //
+    // TODO: Not used. Currently, we never drop relations. The parsing
+    // of relation drops in COMMIT/ABORT records has not been
+    // implemented. We should also have a mechanism to remove
+    // "orphaned" relfiles, if the compute node crashes before writing
+    // the COMMIT/ABORT record.
+    Unlink,
+}
+
+const fn relation_size_key(timelineid: ZTimelineId, rel: RelTag) -> ObjectKey {
+    ObjectKey {
+        timeline: timelineid,
+        buf_tag: BufferTag {
+            rel: rel,
+            blknum: u32::MAX,
+        },
+    }
+}
+
+///
+/// In addition to those per-page and per-relation entries, we also
+/// store a little metadata blob for each timeline. It is stored using
+/// STORAGE_SPECIAL_FORKNUM.
+///
+#[derive(Debug, Clone)]
+#[derive(Serialize, Deserialize)]
+pub struct MetadataEntry {
+    last_valid_lsn: Lsn,
+    last_record_lsn: Lsn,
+    ancestor_timeline: Option<ZTimelineId>,
+    ancestor_lsn: Lsn,
+}
+
+const fn timeline_metadata_key(timelineid: ZTimelineId) -> ObjectKey {
+    ObjectKey {
+        timeline: timelineid,
+        buf_tag: BufferTag {
+            rel: RelTag {
+                forknum: pg_constants::ROCKSDB_SPECIAL_FORKNUM,
+                spcnode: 0,
+                dbnode: 0,
+                relnode: 0,
+            },
+            blknum: 0,
+        },
+    }
+}
+
+
+///
+/// Iterator for `object_versions`. Returns all page versions of a given block, in
+/// reverse LSN order. This implements the traversal of ancestor timelines. If
+/// a page isn't found in the most recent timeline, this iterates to the parent,
+/// until a page version is found.
+///
+struct ObjectVersionIter<'a> {
+    obj_store: &'a dyn ObjectStore,
+
+    buf_tag: BufferTag,
+
+    current_iter: Box<dyn ObjectVersionIterator + 'a>,
+    ancestor_timeline: Option<ZTimelineId>,
+    ancestor_lsn: Lsn,
+}
```

Comment:
`current_iter` could probably use a comment, because there are some confusingly similar names, and that trait is empty, leaving it unclear what's going on.